### PR TITLE
Remove `DefiningAnchor::Bubble` from MIR validation

### PR DESCRIPTION
**Pull request by compiler-errors**
If we're in the MIR validator, our body is local, and our reveal mode is `Reveal::UserFacing`, then use `DefiningAnchor::Bind` -- otherwise, we should never be constraining opaques, so use `DefiningAnchor::Error` like codegen does.

r? oli-obk

### CODE REVIEW

File: compiler/rustc_const_eval/src/interpret/eval_context.rs [hunk 1]
```diff
@@ -384,7 +385,14 @@ pub(super) fn mir_assign_valid_types<'tcx>(
     // all normal lifetimes are erased, higher-ranked types with their
     // late-bound lifetimes are still around and can lead to type
     // differences.
-    if util::relate_types(tcx, param_env, Variance::Covariant, src.ty, dest.ty) {
+    if util::relate_types(
+        tcx,
+        param_env,
+        DefiningAnchor::Error,
```
Comment (compiler-errors) [inline compiler/rustc_const_eval/src/interpret/eval_context.rs:391]: CTFE is always performed with a `Reveal::All` param-env (or it should, as far as I'm aware?) so it makes sense to use `DefiningAnchor::Error` here.

Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/eval_context.rs:391]: I don't think that's true? We run CTFE for const generics as well and the param-env is different there.
Cc @oli-obk @lcnr

File: compiler/rustc_const_eval/src/util/compare_types.rs [hunk 2]
```diff
@@ -59,19 +60,23 @@ pub fn relate_types<'tcx>(
...
-        if hidden_ty != ty.hidden_type.ty {
-            span_bug!(
-                ty.hidden_type.span,
-                "{}, {}",
-                tcx.type_of(key.def_id).instantiate(tcx, key.args),
-                ty.hidden_type.ty
-            );
+
+    if anchor != DefiningAnchor::Error {
+        // With `Reveal::All`, opaque types get normalized away, with `Reveal::UserFacing`
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/util/compare_types.rs:65]: The relationship between what the `if` is checking (`DefiningAnchor`, whatever that is) and what the comment talks about (`Reveal::UserFacing`/`Reveal::All`) is completely unclear.

Comment (compiler-errors) [reply compiler/rustc_const_eval/src/util/compare_types.rs:65]: Well I didn't touch that comment, but I'll rework it I guess.

File: compiler/rustc_const_eval/src/interpret/eval_context.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_const_eval/src/interpret/eval_context.rs:391]: My justification is wrong here. We're not using a `Reveal::All` param-env, but we *have* passed this body through the `RevealAll` mir-pass.

Not totally clear if that's *sufficient*, but if it isn't, then we're at least throwing constrained opaques onto the floor during this check in CTFE.

Comment (compiler-errors) [reply compiler/rustc_const_eval/src/interpret/eval_context.rs]: We can also approximate the defining anchor from the current stack frame. See newest commit for approach.

**Comment by RalfJung** [comment]
Is there any explanation anywhere on what a `DefiningAnchor` is? The type sadly does not have a doc comment that would help here.

**Pull request closed without merging**